### ISSUE TITLE
Optionally support the Native File System API

### ISSUE THREAD

**Issue by tomayac** [photopea/photopea#2118]
Web apps like [Excalidraw](https://excalidraw.com/) have started to support the [Native File System API](https://wicg.github.io/native-file-system/). Google Chrome Labs maintains a [library](https://github.com/GoogleChromeLabs/browser-nativefs) that makes supporting this API a smooth progressive enhancement, with transparent fallbacks to the legacy `input[type=file]` and `a[download]` operation. Here's the basic API: 

```js
 import {
  fileOpen,
  directoryOpen,
  fileSave,
} from 'https://unpkg.com/browser-nativefs';

(async () => {
  // Open a file.
  const blob = await fileOpen({
    mimeTypes: ['image/*'],
  });

  // Open multiple files.
  const blobs = await fileOpen({
    mimeTypes: ['image/*'],
    multiple: true,
  });

  // Open all files in a directory,
  // recursively including subdirectories.
  const blobsInDirectory = await directoryOpen({
    recursive: true
  });

  // Save a file.
  await fileSave(blob, {
    fileName: 'Untitled.png',
  });
})();
```

**Comment by photopea** [comment]
Hi Thomas, we are planning to use the Native File System API, as soon as there are browsers which support it.

I just see one possible problem. People usually hate it, when a webpage asks for permissions right after you open it. People discard such permissions without even reading what it stands for.

There are about 60 reported issues here from people, who did not give Photopea permissions to access the clipboard (because they canceled it automatically), and these people complain, that they can not copy-paste in Photopea.

At what point should a program ask for the access to a Native Filesystem? If we do it too soon, people close it automatically and then, there will be tens of thousands of people, who will just give up using Photopea, as they are unable to save their work.

**Comment by tomayac** [comment]
Have you tried the experience on Excalidraw? It's on-demand and very contextual. Try drawing a diagram, then save it, then make further edits, then save again. Then close everything and re-open the file from disk. 

> [W]e are planning to use the Native File System API, as soon as there are browsers which support it.

This is great news. Note that with the browser-nativefs library, the API is really a progressive enhancement and you only load the code path that's needed. Try Excalidraw in Safari and in Chrome for the full picture.

**Comment by photopea** [comment]
I think I do not need such library. I was not able to save a document in Excalidraw without a Save dialog in between, so I guess the API is not available in the latest Chrome.

**Comment by tomayac** [comment]
> I think I do not need such library.

It's an offer, since the API shape is expected to change during the origin trial phase. The library abstracts away these changes to minimize developer work.

> I was not able to save a document in Excalidraw without a Save dialog in between, so I guess the API is not available in the latest Chrome.

I think there's a misunderstanding. Unlike `a[download]` where file saves automatically land in the user's default Downloads/ folder and where files can't be "over-saved", the Native File System API lets the user save a file to anywhere they want and to "over-save" arbitrarily.

**Comment by photopea** [comment]
@tomayac By the way, do you think Google Drive could provide the tiny JS library, which (after some authentication) would allow us to work with Google Drive with an API, that is 100% identical to the Native File System API?

**Comment by tomayac** [comment]
I think the closest to what you describe is something like in their [quickstart tutorial](https://developers.google.com/drive/api/v3/quickstart/js). I doubt that 100% compatibility with the Native File System API is achievable.

**Comment by photopea** [comment]
@tomayac Do you think you could make a demo for the Asynchronous Clipboard API with images?  https://web.dev/image-support-for-async-clipboard/

The one on JSFiddle does not work.

**Comment by photopea** [comment]
@tomayac Also, do you think you could make this feature consistent across all Chrome browsers? In Chrome for Android, when pasting, I get a message "No valid data on clipboard".

Or is there any way to detect, if this feature is available or not in a current Chrome browser?

**Comment by tomayac** [comment]
I just tried the demo successfully on desktop macOS Chrome 84, it even works (the pasting part) on desktop Safari Technology Preview 108. I did, however, encounter an issue on Android Chrome 83 and filed an issue: https://bugs.chromium.org/p/chromium/issues/detail?id=1094798 (this error doesn't occur on Android Chrome 85 at least). Could you please try again and let me know the browser version and OS where you still see an error?

**Comment by photopea** [comment]
@tomayac The demo https://i.imgur.com/aJniNMx.png does not work in a current stable Chrome (83). Please, try it in Chrome which people use, without any flags.

**Comment by tomayac** [comment]
I just tried it on Chrome 83 on Windows 10, without any flags, and it worked. What OS are you on?

**Comment by photopea** [comment]
I use Chrome 83 on Windows 10. The demo is in a sandboxed iframe.

As I load the link, the console shows warnings:

    Unrecognized feature: 'vr'.
    Unrecognized feature: 'clipboard'.
    Unrecognized feature: 'clipboard-read'.
    Unrecognized feature: 'clipboard-write'.

When I press Copy, I see NotAllowedError: Disabled in this document by Feature Policy.

**Comment by tomayac** [comment]
Oh, this is a clarification that was missing. For being able to run in sandboxed iframes, we're dependent on https://github.com/w3c/webappsec-feature-policy/issues/322, which is still open.

**Comment by dentistformyeye** [comment]
Would the current method of opening and saving remain available for users who choose to disable the Native File System API?

**Comment by tomayac** [comment]
> Would the current method of opening and saving remain available for users who choose to disable the Native File System API?

I'm not sure there is or will be a way to disable the API, but let's just assume there is. In this case (and in the case of a browser that doesn't implement the API to begin with), the API's feature detection `if ('showOpenFilePicker' in window) { /* Supported */}` would fail, and the alternative implementation would be used. Our library, [browser-nativefs](https://github.com/GoogleChromeLabs/browser-nativefs), is built to support this approach.

**Comment by photopea** [comment]
We are still waiting for some browsers to support this API.

**Comment by tomayac** [comment]
You could use it as a progressive enhancement today.

**Comment by photopea** [comment]
What is a progressive enhancement? To me, it makes no sense to do something nobody can use.

**Comment by tomayac** [comment]
"Nobody" in this case is everyone using Chromium >=86. The API has [launched](https://web.dev/file-system-access/#status:~:text=Chrome%2086) now. It's usable today with [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement) paired with [feature detection](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Feature_detection).

**Comment by photopea** [comment]
According to my Google Analytics, only 0.03% of users have Chromium >=86, that is what I mean by "nobody".

**Comment by tomayac** [comment]
It's definitely a moving target that will grow rapidly once 86 becomes the stable version.

**Comment by dentistformyeye** [comment]
I'm currently using a Chrome flag to disable the API, but the flag may get removed after a few updates. The devs of my browser are planning to disable the API, although I'm not sure whether it'll be permanent.

**Comment by photopea** [comment]
@tomayac I would like to implement a File System API into Photopea. Currently, users can open files in Photopea with File - Open, or by drag-and-dropping.

I see, that with the File System API, I should always start by calling window.showOpenFilePicker().

With Drag And Drop, my webpage is listening to the "drop" event. 

Then, I access event.dataTransfer.files, which is a FileList object. I loop through it to access File objects.

For each File, I create var fr = new FileReader() object, and call fr.readAsArrayBuffer(file);  I handle the files from <input type="file"> the same way.

Is it possible to get a FileSystemFileHandle from a File received by Drag and Drop?

**Comment by tomayac** [comment]
This is possible. Please see this [example](https://wicg.github.io/file-system-access/#draganddrop-example).

**Comment by photopea** [comment]
@tomayac I implemented the API. If you open a PSD file from a computer (File - Open, or drag-and-drop), the File - Save option will be available, which will rewrite the original PSD in a computer.

When a user creates a document from scratch in Photopea, and saves it with File - Save as PSD, I would like the File - Save option to become available after that.

The window.showSaveFilePicker() does not allow specifying a default file name. I would like to use the name of a document in Photopea, but now, the user would have to enter it manually . That is why I did not implement it yet.

I see you have reported it, are there any news?

https://github.com/WICG/file-system-access/issues/80
https://bugs.chromium.org/p/chromium/issues/detail?id=1145102

BTW. have you thought about storing "file system permissions" for a webapp forever, just like what we have with Clipboard permissions, so that the user does not have to confirm it each time? I know I can remember FileHandles, but there still will be a dialog for each new file / directory.

Also, you have a mistake in your article: https://web.dev/file-system-access/  You define "const options = ...", but then, you call "window.showSaveFilePicker(opts);"

**Comment by tomayac** [comment]
> @tomayac I implemented the API. If you open a PSD file from a computer (File - Open, or drag-and-drop), the File - Save option will be available, which will rewrite the original PSD in a computer.

This is brilliant, thanks for implementing this! Are you planning on extending support for other file types (I tested with a PNG, but it only worked for PSD)?
 
> When a user creates a document from scratch in Photopea, and saves it with File - Save as PSD, I would like the File - Save option to become available after that (to rewrite the saved file).

This is possible. You get the file handle back from the save dialog. 
 
> The window.showSaveFilePicker() does not allow specifying a default file name. I would like to use the name of a document in Photopea, but now, the user would have to enter it manually . That is why I did not implement it yet.
> 
> I see you have reported it, are there any news?
> 
> [WICG/file-system-access#80](https://github.com/WICG/file-system-access/issues/80)
> https://bugs.chromium.org/p/chromium/issues/detail?id=1145102

No news yet, but I have assigned an owner manually now, so hopefully they see it. It's been a busy week for our US colleagues…

> BTW. have you thought about storing "file system permissions" for a webapp forever, just like what we have with Clipboard permissions, so that the user does not have to confirm it each time? I know I can remember FileHandles, but there still will be a dialog for each new file / directory.

Storing the handles is currently the best you can do. We also have an [issue](https://github.com/WICG/file-system-access/issues/184) where someone else requests this, so it might be a good idea to chime in there. 

> Also, you have a mistake in your article: https://web.dev/file-system-access/ You define "const options = ...", but then, you call "window.showSaveFilePicker(opts);"

Thank you, I have opened a [PR](https://github.com/GoogleChrome/web.dev/pull/4201) to fix this.

**Comment by photopea** [comment]
@tomayac also, the File System API does not work if a webpage is inserted into a different domain with an iframe https://github.com/photopea/photopea/issues/2596

**Comment by dentistformyeye** [comment]
My disabling the #native-file-system-api flag in chrome://flags makes Photopea crash when I try File > Open or drag-and-drop. [This demo](https://browser-nativefs.glitch.me/) that @tomayac mentioned also doesn't work.
![image](https://user-images.githubusercontent.com/51429646/98691945-93d2ae80-233c-11eb-9ecc-4cda1708d3fb.png)

Fortunately, copy-paste still works, so I have a workaround for PNGs and JPGs. I don't want to make Photopea too big, but I would appreciate a way to toggle to the old method for opening files.

**Comment by photopea** [comment]
@dentistformyeye Why do you modify flags? I think it is be obvious, that if you modify the standard behaviour of a browser (change flags, add extensions, etc), some websites might stop working.

If you really want to use browser this way, I guess you should ask Google to do something with it. Or use another browser.

**Comment by photopea** [comment]
@tomayac BTW. do you know what is the support for File System API in other Chromium browsers?

E.g. in Yandex browser, when you call window.showOpenFilePicker(), it reloads the webpage.

**Comment by tomayac** [comment]
I investigated this issue in Yandex Browser v20.11.0.918 (64-bit) and boiled it down to Yandex exposing the API despite it still being gated behind the `browser://flags/#native-file-system-api` flag in their implementation. Judging from the available flags, it seems like they are based on Chromium 86 with the present release, but seem to have made some adjustments compared to Chrome. You should probably [report feedback](https://browser.yandex.com/feedback/?session_id=20&description=&categoryTag=&feedbackChannel=release&ui=0EC1C596-6170-4005-97B0-8BFD33E6156A&pageUrl=browser%3A%2F%2Fhelp&request_screenshot=false) to the Yandex team.

**Comment by tomayac** [comment]
If you toggle the flag, the API works as expected by the way.

**Comment by photopea** [comment]
@tomayac I am getting reports from many schools, that they can not open any files in Photopea on Chromebooks. My guess is, that it is caused by some of many browser extensions (adblocks etc.), which some "smart people" pre-install into Chromebooks to "protect" users. Is there a way I can detect, that such extension is installed. Teachers say they can not disable these extensions.

Is there a way I can detect, that any of such extensions is present?

A possible solution - could you remove the ability to install extensions into a browser completely? I spent about an hour every day helping people with issues in Photopea, that are caused by browser extensions (and are solved after disabling extensions). 

I usually tell them to try Photopea in Incongito mode and they do not reply after that. I think there are thousands of people who open the Incognito mode each time they want to use Photopea, because it works there. They have no idea what a browser extension is.

I think 90% of people are not aware of all extensions they have in their browser, and 100% of people do not know what exactly their extensions do.

**Comment by tomayac** [comment]
My suggestion would be to listen for errors at the global level through [`window.onerror`](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror#window.onerror) and report those back to a reporting backend. Like that you will get a feel for the errors that happen in the wild, but not on your machine. 

In addition to that, I would propose you take a look at the [Reporting API](https://developers.google.com/web/updates/2018/09/reportingapi), which allows you to be informed of browser interventions, CSP violations, and other issues. 

In combination, this should give you a pretty holistic overview of real world errors your users encounter.

For your concrete questions: extensions can't be detected easily, but one to detect ad blockers is to create honey pot "ads" that would for sure trigger an ad blocker and where you check if the "ad" made its way unaltered into the DOM.

Admins can disable (certain) extensions for managed Chrome OS fleets. The details are up to the administrators, though.

**Comment by photopea** [comment]
@tomayac by the way, will the Local Font Access API be available any time soon? I found this article of yours: https://web.dev/local-fonts/ .

People have to drag-and-drop OTF and TTF files into Photopea, to load their own fonts. By default, people on Windows can not use Arial in Photopea, people on Macs can not use Helvetica font. These fonts come as a part of the OS, and people are surprised, because they can use these fonts in all other apps.

**Comment by tomayac** [comment]
> @tomayac by the way, will the Local Font Access API be available any time soon? I found this article of yours: https://web.dev/local-fonts/ .
> 
> People have to drag-and-drop OTF and TTF files into Photopea, to load their own fonts. By default, people on Windows can not use Arial in Photopea, people on Macs can not use Helvetica font. These fonts come as a part of the OS, and people are surprised, because they can use these fonts in all other apps.

The API is still in flux, see for example this recently [merged PR](https://github.com/WICG/local-font-access/pull/63/files) that talks about adding filters to the `navigator.fonts.query()` method. You can track (i.e., star) the [Chromium bug](https://crbug.com/535764) to stay up-to-date. We definitely want to enable use cases like Photopea's!

**Comment by photopea** [comment]
@tomayac btw. do not create too many tools for working with local fonts in your API. It is enough to only allow getting a list of fonts (postScriptNames), and then, getting a binary file (OTF / TTF) for each postScriptName.

**Comment by tomayac** [comment]
We're balancing a lot of partners' needs, it's good to hear your perspective, too.

**Comment by photopea** [comment]
@tomayac  Do you know what is the problem with this feature in Opera?

File - Save as PSD in Photopea uses window.showSaveFilePicker(), if it is present on the window object.

Now, when you press File - Save as PSD in Photopea in Opera, it creates a file with the length 0. Several people complained to me, that they lost heir work this way. I could not get in touch with anyone from Opera. Could you tell them to look into it?

**Comment by photopea** [comment]
@tomayac I found a serious bug with showSaveFilePicker(), and I need to remove it from Photopea completely. Do you think you could fix it? https://bugs.chromium.org/p/chromium/issues/detail?id=1232301

**Comment by tomayac** [comment]
> @tomayac Do you know what is the problem with this feature in Opera?
> 
> File - Save as PSD in Photopea uses window.showSaveFilePicker(), if it is present on the window object.
> 
> Now, when you press File - Save as PSD in Photopea in Opera, it creates a file with the length 0. Several people complained to me, that they lost heir work this way. I could not get in touch with anyone from Opera. Could you tell them to look into it?

Ouch, this is [a bug](https://opera.atlassian.net/servicedesk/customer/portal/9/BS-15494?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJxc2giOiJmYmU1NWQ2YjNkODUwNGIyOTRhOTM1ZmFkNDY3M2I1MmQ0ZDI2ODA1NDhlNDhhOWFjZTA2ZWUzY2VlOGM5ZDYyIiwiaXNzIjoic2VydmljZWRlc2stand0LXRva2VuLWlzc3VlciIsImNvbnRleHQiOnsidXNlciI6IjMwNzQ4IiwiaXNzdWUiOiJCUy0xNTQ5NCJ9LCJleHAiOjE2Mjk0NDYyODgsImlhdCI6MTYyNzAyNzA4OH0.Jqcvo5fiqRpMQFdQREN67ajS1ZOhz7JD-GS1qLA4-JE&sda_source=notification-email) with Opera's implementation. I have reported it to them, it's tracked as `BS-15494 Saving files with the File System Access API fails`.

For your users: it looks like the browser fails to switch the temporary `.crswap` file for the final file. Tell your users to look for a `.crswap` file to recover their work.

<img width="537" alt="Screen Shot 2021-07-23 at 09 55 03" src="https://user-images.githubusercontent.com/145676/126753483-cc65db9c-aa35-47e1-85be-6b1e7ecbbe9b.png">

FWIW, I also noticed [a bug](https://opera.atlassian.net/servicedesk/customer/portal/9/BS-15493?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJxc2giOiI0NGRlY2ZlMTVmNTk0MTlmZjJhM2UyOWI4ZGQ1OTg3Yjc4Y2I5MzY3MzAzMGU4NzRhZWVmZGM5NTg4MzMwYjI4IiwiaXNzIjoic2VydmljZWRlc2stand0LXRva2VuLWlzc3VlciIsImNvbnRleHQiOnsidXNlciI6IjMwNzQ4IiwiaXNzdWUiOiJCUy0xNTQ5MyJ9LCJleHAiOjE2Mjk0NDU5NjIsImlhdCI6MTYyNzAyNjc2Mn0.MHV9b71tv1E-9rfCmS10CHqAh9EMIpV_ktpsQLGCZVI&sda_source=notification-email) with their `showDirectoryPicker()` implementation. It's tracked as ``BS-15493 'showDirectoryPicker()' throws a 'Uncaught DOMException: The user aborted a request.'``.

**Comment by tomayac** [comment]
> @tomayac I found a serious bug with showSaveFilePicker(), and I need to remove it from Photopea completely. Do you think you could fix it? https://bugs.chromium.org/p/chromium/issues/detail?id=1232301

This is a duplicate of [crbug.com/1193489](https://bugs.chromium.org/p/chromium/issues/detail?id=1193489) (I ran into the same issue), which has a solution outlined in [comment #1](https://bugs.chromium.org/p/chromium/issues/detail?id=1193489#c1): 

_"Is there some reason "file save prep work" has to be done in the time in between clicking the button and showing the file dialog? I.e. why not show the file dialog first, and then do the work that is needed to actually save the file?"_

So essentially get the file handle first, and then write into it.

**Comment by photopea** [comment]
Oh, you are right, I did not realize it can be done this way. I did it this way. Thanks a lot! :)

But there is still a difference between the two bug reports. You do setTimeout() - "passive waiting", while in my case, there is actual CPU work - "active waiting". When there is an active work in between, I think the browser should consider it to be done "during a gesture".

**Comment by tomayac** [comment]
> Oh, you are right, I did not realize it can be done this way. I did it this way. Thanks a lot! :)

Happy you found a solution :-) 
 
> But there is still a difference between the two bug reports. You do setTimeout() - "passive waiting", while in my case, there is actual CPU work - "active waiting". When there is an active work in between, I think the browser should consider it to be done "during a gesture".

I agree with Marijn here: _" Showing a dialog an arbitrary time after user activation doesn't seem like it would be a great user interface."_ In the worst case your CPU task could take way longer than the time a regular user would expect a dialog to show.

**Comment by photopea** [comment]
@tomayac  There is one more problem. When you right-click a JPG image in Windows 10 File Explorer, it can be pasted in Imgur.com , but not in Photopea.com (error: "No valid data on clipboard.") . Photopea uses the new Clipboard API, while Imgur probably does not.

Is it possible, that the "old way" of pasting is more capable than the new API?

**Comment by tomayac** [comment]
You can do this, please see this [demo](https://coal-serious-enigmosaurus.glitch.me/):

```js
document.addEventListener("paste", e => {
  if (!e.clipboardData.files.length) {
    return;
  }
  [file] = e.clipboardData.files;
  console.log(file);
  const image = document.createElement("img");
  image.src = URL.createObjectURL(file);
  document.body.append(image);
});

```

**Comment by photopea** [comment]
So should I use the "paste" event for pasting, or navigator.clipboard.read() ? I want to be able to initiate pasting from Photopea (with Edit - Paste), and I want it to work the same as "native pasting", because otherwise, people would be confused, why it works differently.

**Comment by tomayac** [comment]
It works either way for images copied via the context menu, by listening to the `paste` event or by having a dedicated **Paste** button. I have updated the [demo](https://coal-serious-enigmosaurus.glitch.me/) accordingly:

```js
document.querySelector("button").addEventListener("click", async () => {
  try {
    const items = await navigator.clipboard.read();
    if (!items) {
      return;
    }
    console.log(items);
    for (const item of items) {
      for (const type of item.types) {        
        if (type.startsWith("image/")) {
          const blob = await item.getType(type);
          const image = document.createElement("img");
          image.src = URL.createObjectURL(blob);
          document.body.append(image);
          return;
        }
      }
    }
  } catch (err) {
    console.warn(err.name, err.message);
  }
});
```

Just note that the API _re-encodes_ the `.jpg` image to `.png`, but it's a lossless operation obviously. This is also the reason why `.jpg` images copied from, for example, the file explorer won't work. Star [crbug.com/954390](https://crbug.com/954390) for JPEG clipboard support.

**Comment by photopea** [comment]
When I copy a JPG or a PNG from a file explorer, it can not be pasted in your demo.

So I just want to make sure, that you are aware of a fact, that the clipboard API is less capable than the "paste" event. Do you consider it to be a bug? Do you think the Clipboard API should be equally capable as pasting an image with the "paste" event?

**Comment by tomayac** [comment]
Yes, I am aware that `paste` is more powerful than `navigator.clipboard.read()` at the moment, and yes, I think both APIs should be equally capable. Work to make this happen is underway in [crbug.com/954390](https://crbug.com/954390), as pointed out earlier in the thread.

**Comment by photopea** [comment]
@tomayac ok, I see, thanks! :)

BTW. I noticed, that right now, showSaveFilePicker() requires some "gesture" in the UI. I need to call it after confirming the window.confirm() dialog, which does not work now, so switched to the old version, without showSaveFilePicker() .

I know, that you do all these things in order to "protect" users from "bad" websites. Just like re-encoding images in a clipboard, or not allowing websites to detect the Ctrl+T shortcut. I think you are restricting websites too much, without giving them a chance to ever reach capabilities of native apps.

**Comment by tomayac** [comment]
Correct, we require a user gesture that cannot be passed on from a `confirm()` dialog.

Some interesting news:
- [WebCodecs](https://web.dev/webcodecs/) are shipping in Chrome 94 and may help with the re-encoding.
- The `ctrl+t` shortcut can be intercepted in fullscreen mode by requesting a [keyboard lock](https://web.dev/keyboard-lock/).

**Comment by photopea** [comment]
@tomayac thanks! I will use the this for downloading files, it does not require a gesture.

    var a = document.createElement( "a" );
    a.href = window.URL.createObjectURL( new Blob([myArrayBuffer]) );
    a.download = "file.psd";
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);

I know that you allow Ctrl+T in fullscreen, but I would like to offer a consistent experience in all cases. Telling users to "switch to fullscreen to unlock more features" sounds really weird to me, they would not take it seriously.

**Comment by tomayac** [comment]
This a bit of a step back, since downloading is inferior to actually saving and over-saving. Can we take a step back and can I ask you why the `confirm()` dialog is needed? I didn't see one when I just tried the app. Are there special conditions that need to be met?

As a side note, there are [long-term plans](https://groups.google.com/a/chromium.org/g/blink-dev/c/hTOXiBj3D6A/m/QnvjjvCSAwAJ) to remove `alert()` and `confirm()` and friends from the platform.

**Comment by photopea** [comment]
Sometimes, users run a script in Photopea, which finishes with saving data to a computer (a ZIP file).

Why do you require a gesture, if the user will see the Saving window anyway? So websites can not write to a hard drive without the user confirming it first.

**Comment by tomayac** [comment]
I recall we talked about this [before](https://github.com/photopea/photopea/issues/2118#issuecomment-885475834). The trick is to get the handle first, and only then run the potentially long-running operation  like zipping.

(By the way, are you aware of [`CompressionStream`](https://developer.mozilla.org/en-US/docs/Web/API/CompressionStream)?)

**Comment by photopea** [comment]
@tomayac Do you think Opera will ever fix it? There are people who lost all their saved work, because they used Opera, see https://github.com/photopea/photopea/issues/3748

**Comment by tomayac** [comment]
I unfortunately don't know anyone at Opera anymore, so can't do more than complain on their [reporting portal](https://opera.atlassian.net/servicedesk/customer/portal/9), which is what I did.

**Comment by photopea** [comment]
@tomayac  It seems like Opera has fixed it in the latest release.

BTW. many schools embed Photopea into their websites using the <iframe>, so that students can use it there. Why do you block showOpenFilePicker() when a website runs in an iframe on another domain?

**Comment by tomayac** [comment]
> @tomayac It seems like Opera has fixed it in the latest release.

That's welcome news!

> BTW. many schools embed Photopea into their websites using the <iframe>, so that students can use it there. Why do you block showOpenFilePicker() when a website runs in an iframe on another domain?

I have pinged https://github.com/WICG/file-system-access/issues/245#issuecomment-926755747.

**Comment by photopea** [comment]
@tomayac Hi! I want to show the user the content of some folder in their computer. I obtain a "FileSystemDirectoryHandle" and print the files without any problem.

Next time the user opens Photopea, I want to show them the content of the folder, which the user opened last time, without asking for permissions again, etc. I saved the "FileSystemDirectoryHandle" into the IndexedDB last time, and I can load it now.

However, now, doing "for await (var entry of folder.values()) ..."  throws an error now:  "Uncaught (in promise) DOMException: The request is not allowed by the user agent or the platform in the current context." 

Is it possible to obtain permissions to access a file system, so that these permissions last even after reopening a website? Or do I have to bother a user with popups and requests every time?

**Comment by tomayac** [comment]
Unfortunately for now you need to re-prompt. This is a situation we want to improve, though. Please subscribe to https://github.com/WICG/file-system-access/issues/297 for updates on this.

**Comment by photopea** [comment]
@tomayac thanks a lot! BTW. is there a way to rename a file or a folder? I know I could do it by loading a file into a memory, creating a copy with a new name, and deleting the old one. But I wish it could be done without changing the "lastModified".

**Comment by tomayac** [comment]
Try in 98 (I tested in 98.0.4726.0 (Official Build) canary (arm64)):

```js
[file] = await showOpenFilePicker()
// [FileSystemFileHandle]
await file.queryPermission({mode:'readwrite'})
// 'prompt'
await file.requestPermission({mode:'readwrite'})
// 'granted'
await file.rename('foo')
// undefined
file
// FileSystemFileHandle {kind: 'file', name: 'foo'}
```

**Comment by photopea** [comment]
Ok, I will wait until the stable chrome supports it, thanks :)

**Comment by photopea** [comment]
@tomayac I am saving a file this way:

    handle.createWritable.().then(function(e) {e.write(buff); return e;}).then(
        function(e) {  e.close();   showMessage("Saved!");   }  );

I noticed, that after I show "Saved!" to the user, and the user closes Photopea immediately, there is only a zero-Byte file on a hard drive. It is needed to wait 3 - 6 seconds before closing a website, to be sure a file is saved correctly.

Am I saving it a wrong way, or is it a bug in Chrome?

**Comment by tomayac** [comment]
@photopea This happens when the stream hasn't been closed. 

I suggest you move to `async`/`await`:

```js
const writable = await fileHandle.createWritable();
// Write the contents of the file to the stream.
await writable.write(contents);
// Close the file and write the contents to disk.
await writable.close();
```

**Comment by photopea** [comment]
@photopea but my code does the same thing, so it is identical to your code. The bug will happen in your code, too.

Or should I wait until the "close()" has finished, i.e. add another ".then()" ?

**Comment by photopea** [comment]
@tomayac but my code does the same thing (syntactically), so it is identical to your code. The bug will happen in your code, too.

I noticed, that if I redirect to another website after calling writable.close();  it is still saved correctly. But if I close a browser tab after calling writable.close();  I end up with a zero-byte file.

**Comment by tomayac** [comment]
Tiny difference: You don't `await` the `close()` call before showing the "saved!" message.

**Comment by photopea** [comment]
@tomayac BTW. I think you should discourage people from using async/await when working with the Native File System API.

E.g. I have a folder with 344 files inside. Looping through all FileSystemDirectoryHandle.values() takes 10 ms. But I need to find out a size of each file, so I have to call FileSystemFileHandle.getFile().

If I call await ...getFile() inside a loop, it takes 1000 ms.

If I call ...getFile().then(...), to make them "load in parallel", all files are loaded in 30 ms (but it is more code, you need to track the number of finished files).

**Comment by tomayac** [comment]
Thanks for the proposal to document this. https://github.com/GoogleChrome/web.dev/pull/7005

**Comment by photopea** [comment]
@tomayac I want to batch-process hundreds of files form a users hard drive. I have a "FileSystemDirectoryHandle" from the user, with read/write permissons.

I would like to create 100 new files in that directory, with new names (which do not exist yet). Can it be done with the current API now?

**Comment by photopea** [comment]
@tomayac BTW. let me show you what I made :)  After you go to https://www.photopea.com/ , there are "storages" on the left side. Basically, it is a file explorer for different "storages", one of them is a local file system through the Native File System API.

**Comment by tomayac** [comment]
Creating 100s of files is definitely possible: https://momentous-iridescent-soup.glitch.me/

```js
const button = document.querySelector("button");

button.addEventListener("click", async () => {
  const handle = await showDirectoryPicker();
  for (let i = 0; i < 100; i++) {
    handle.getFileHandle(`file-${i}.txt`, { create: true });
  }
});
```

**Comment by tomayac** [comment]
Just made some noise about your new feature: https://twitter.com/tomayac/status/1470329245553549313. Really love where this is going!

**Comment by photopea** [comment]
@tomayac thanks a lot! BTW. do you know anything about the Google Drive API? I implemented it as another storage, next to the file system, but Goolge does not allow many users to connect it, saying that the app (Photopea) is dangerous, see https://github.com/photopea/photopea/issues/4114 . Can you connect to your Google Drive in Photopea?

**Comment by tomayac** [comment]
I responded over on #4114. I could grant access via "Advanced", but it's really not the recommended way.

**Comment by photopea** [comment]
There is another problem with the Native File System API.

People work with large files in Photopea. Let's save it takes 5 seconds to serialize a document into a 200 MB binary file, and 5 seconds to write it to a hard drive. When the user presses File - Save, I can not serialize the document (because "showSaveFilePicker" requires a gesture). After I show the save file picker, and the user hits save, there message "Saving ..." is displayed for 10 seconds, so they must wait 10 seconds.

Many people hit Save and close Photopea immediately (they expect it to be saved), see here, lots of people lose their work: https://www.reddit.com/r/photopea/comments/s1p36o/photopea_deleted_saves_from_my_hardrive/

How should we solve it? Could you make Chrome not close a website, until writing to a hard drive is finished? But they might close it during the serialization, before saving :(

**Comment by tomayac** [comment]
I guess this is what the [`onbeforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload) is good for. You could add an event listener for this event until the file is saved for good, and remove it afterwards.

**Comment by photopea** [comment]
@tomayac that is what I am doing. But the best I can do is to make the browser show a message "Are you sure you want to leave this page?". But users click "Yes" anyway, and their work is lost.

**Comment by tomayac** [comment]
That's great! Probably a more aggressive "saving in progress" information would help people understand the app is busy. Maybe as aggressive as blurring the entire app and having an info modal.

**Comment by photopea** [comment]
Hi @tomayac , did you do any changes to the way the Native File System API works in Chrome 98?

It seems like now, if I close the Photopea tab before the saving finishes, it still finishes saving (and does not produce a 0-Byte file). Or is it an error on my side?

**Comment by photopea** [comment]
There is one more problem with the API. 

I call "showSaveFilePicker" and the user chooses an existing file (it asks you if you want to replace that file, you confirm it). Then, I get a file handle. But I never call "createWritable", "write", "close". In that case, the content of the old file is erased, it is set to 0 Bytes.

Could you confirm, that you are aware of this behaviour, and that it is wrong and should be fixed? Then, I will report it as a bug.

Users often choose to replace an old file in the computer when saving. But if Photopea e.g. runs out of RAM while generating a binary file, it crashes, so their current work is lost, and the old version of their work is deleted by Chrome, so it is lost also. (And of course, I can not generate a file before calling showSaveFilePicker, because you require a gesture, and there would be a delay).

**Comment by tomayac** [comment]
> It seems like now, if I close the Photopea tab before the saving finishes, it still finishes saving (and does not produce a 0-Byte file). Or is it an error on my side?

It sounds like this is the behavior the user expects. There are several mentions of "file system" in the [changelog](https://chromium.googlesource.com/chromium/src/+log/97.0.4692.99..98.0.4758.82?pretty=fuller&n=10000) indeed.

> I call "showSaveFilePicker" and the user chooses an existing file (it asks you if you want to replace that file, you confirm it). Then, I get a file handle. But I never call "createWritable", "write", "close". In that case, the content of the old file is erased, it is set to 0 Bytes.

This is working as [specified](https://wicg.github.io/file-system-access/#dom-window-showsavefilepicker), specifically, [step 7](https://wicg.github.io/file-system-access/#:~:text=Set%20entry%E2%80%99s%20binary%20data%20to%20an%20empty%20byte%20sequence.):

> 7. Set entry’s [binary data](https://wicg.github.io/file-system-access/#file-entry-binary-data) to an empty [byte sequence](https://infra.spec.whatwg.org/#byte-sequence).

**Comment by photopea** [comment]
@tomayac Do you think there is a chance to change it? I think it is wrong to delete a file, if the new content is never written.

Can you see the problem in my case? I need to create a binary file, which takes time and memory, and can be killed by a web browser. So if I do it after the user clicks "Save", before calling "showSaveFilePicker", the "showSaveFilePicker" does not receive a gesture. If I call it after "showSaveFilePicker", before "createWritable", and Chrome kills Photopea, the users file gets deleted.

**Comment by photopea** [comment]
@tomayac hey, do you know when will Chrome support the "file_handlers"? https://web.dev/file-handling/ You wrote that article almost two years ago, but it is still not available. It is implemented in Photopea already.

7.3% of our users start Photopea with an icon from their homescreen. If "file_handlers" were supported, it would be a nice "gift" to our users.

**Comment by tomayac** [comment]
@photopea It's scheduled for launch in M-102 [[source](https://bugs.chromium.org/p/chromium/issues/detail?id=829689#c142)].
